Sources/AWSLambdaRuntimeCore/ControlPlaneResponseDecoder.swift

```diff
@@ -0,0 +1,525 @@
+            case .headerEnd:
+                self.state = .waitingForBody(head)
+                return .decoded(head)
+
+            case .contentLength(let length):
+                head.contentLength = length // TODO: This can crash
+
+            case .contentType:
+                break // switch
+
+            case .requestID(let requestID):
+                head.requestID = requestID
+
+            case .traceID(let traceID):
+                head.traceID = traceID
+
+            case .functionARN(let arn):
+                head.invokedFunctionARN = arn
+
+            case .cognitoIdentity(let cognitoIdentity):
+                head.cognitoIdentity = cognitoIdentity
+
+            case .deadlineMS(let deadline):
+                head.deadlineInMillisSinceEpoch = deadline
+
+            case .ignore:
+                break // switch
+            }
+        }
+    }
+
+    enum BodyEncoding {
+        case chunked
+        case plain(length: Int)
+        case none
+    }
+
+    private mutating func decodeBody(from buffer: inout ByteBuffer) throws -> DecodeResult<ByteBuffer?> {
+        switch self.state {
+        case .waitingForBody(let partialHead):
+            switch partialHead.contentLength {
+            case .none:
+                return .decoded(nil)
+            case .some(let length):
+                if let slice = buffer.readSlice(length: length) {
+                    self.state = .waitingForNewResponse
+                    return .decoded(slice)
+                }
+                return .needMoreData
+            }
+
+        case .waitingForNewResponse, .parsingHead, .receivingBody:
+            preconditionFailure("Invalid state: \(self.state)")
+        }
+    }
+
+    private mutating func decodeResponse(head: PartialHead, body: ByteBuffer?) throws -> ControlPlaneResponse {
+        switch head.statusCode {
+        case 200:
+            guard let body = body else {
+                preconditionFailure("TODO: implement")
+            }
+            return .next(try Invocation(head: head), body)
+        case 202:
+            return .accepted
+        case 400 ..< 600:
+            preconditionFailure("TODO: implement")
+
+        default:
+            throw LambdaRuntimeError.unexpectedStatusCode
+        }
+    }
+
+    mutating func decodeStatusLine(from buffer: inout ByteBuffer) throws -> Int {
+        guard buffer.readableBytes >= 11 else {
+            throw LambdaRuntimeError.responseHeadInvalidStatusLine
+        }
+
+        guard buffer.readString("HTTP/1.1 ") else {
+            throw LambdaRuntimeError.responseHeadInvalidStatusLine
+        }
+
+        let statusAsString = buffer.readString(length: 3)!
+        guard let status = Int(statusAsString) else {
+            throw LambdaRuntimeError.responseHeadInvalidStatusLine
+        }
+
+        return status
+    }
+
+    private mutating func decodeCRLFTerminatedLine(from buffer: inout ByteBuffer) throws -> DecodeResult<ByteBuffer> {
```

Comment:
does nio offer any utils to parse line by line, I thought there was some Decoder for that?